pkg/operator/controllers/machine_controller.go

```diff
@@ -0,0 +1,203 @@
+	clusterapi "github.com/openshift/cluster-api/pkg/client/clientset_generated/clientset"
+	"github.com/operator-framework/operator-sdk/pkg/status"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	aro "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/typed/aro.openshift.io/v1alpha1"
+	_ "github.com/Azure/ARO-RP/pkg/util/scheme"
+)
+
+const (
+	machineSetsNamespace = "openshift-machine-api"
+)
+
+// MachineReconciler reconciles the alertmanager webhook
+type MachineReconciler struct {
+	clustercli      clusterapi.Interface
+	arocli          aroclient.AroV1alpha1Interface
+	log             *logrus.Entry
+	developmentMode bool
+	role            string
+}
+
+func NewMachineReconciler(log *logrus.Entry, clustercli clusterapi.Interface, arocli aroclient.AroV1alpha1Interface, role string) *MachineReconciler {
+	return &MachineReconciler{
+		clustercli:      clustercli,
+		arocli:          arocli,
+		log:             log,
+		role:            role,
+		developmentMode: true, //TODO we need this!!
+	}
+}
+
+func vmSizeIsValid(vmSize api.VMSize, developmentMode bool, isMaster bool) bool {
+	if isMaster {
+		switch vmSize {
+		case api.VMSizeStandardD8sV3,
+			api.VMSizeStandardD16sV3,
+			api.VMSizeStandardD32sV3:
+		default:
+			return false
+		}
+	} else {
+		if developmentMode {
+			if vmSize != api.VMSizeStandardD2sV3 {
+				return false
+			}
+		} else {
+			switch vmSize {
+			case api.VMSizeStandardD4asV4,
+				api.VMSizeStandardD8asV4,
+				api.VMSizeStandardD16asV4,
+				api.VMSizeStandardD32asV4,
+				api.VMSizeStandardD4sV3,
+				api.VMSizeStandardD8sV3,
+				api.VMSizeStandardD16sV3,
+				api.VMSizeStandardD32sV3,
+				api.VMSizeStandardE4sV3,
+				api.VMSizeStandardE8sV3,
+				api.VMSizeStandardE16sV3,
+				api.VMSizeStandardE32sV3,
+				api.VMSizeStandardF4sV2,
+				api.VMSizeStandardF8sV2,
+				api.VMSizeStandardF16sV2,
+				api.VMSizeStandardF32sV2:
+			default:
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (r *MachineReconciler) machineSupported(ctx context.Context, machine *machinev1beta1.Machine) (bool, []string, error) {
+	msgs := []string{}
+	supported := true
+
+	role, roleOK := machine.Labels["machine.openshift.io/cluster-api-machine-role"]
+	if !roleOK {
+		return true, nil, fmt.Errorf("role not set on machine label yet")
+	}
+	isMaster := role == "master"
+
+	machinesetOK := false
+	for _, ref := range machine.GetOwnerReferences() {
+		if ref.Kind == "MachineSet" {
+			machinesetOK = true
+		}
+	}
+	if !machinesetOK && !isMaster {
+		supported = false
+		msgs = append(msgs, fmt.Sprintf("the machine %s does not belong to a machineset", machine.Name))
+	}
```

Comment:
@olga-mir was a bit sceptical whether this would always catch the case of orphan machine.
I might do a machineset list to read the expected number of replicas

Reply:
She is right. I can create object from other clusters and have owner ref to non-existing machineset